### DISCUSSION TITLE
Any ideas of a workspace management feature?

### DISCUSSION THREAD

**Discussion by renyuneyun** [piroor/treestyletab#3505] [Ideas]
## Background / Current problem
There are couple of dedicated workspace management extensions for firefox (and guess the same for chrome/chromium), but they all have the following issues:
- Unaware of tree style tabs (or similar extensions).
- Cannot store to bookmark and thus sync across devices.

In the meantime, TST has the following drawbacks for its "saving sub-tree as bookmark" function:
- Saving and loading must be manually triggered.
- Will not be in sync after loading (e.g. opened a new sub-tab, but this new tab won't be added in the bookmark).

## Proposed feature
It would be really nice if these two can be combined, resulting in:

- Open tabs with tree-style, in a "workspace"
- The workspace is automatically saved in bookmarks, not to external cloud servers
- Whenever a new tab is open in this workspace, it gets added to the bookmark as well
- Workspaces can be closed in one go, and restored in the future
- [Optional] Other machines can see and manage the same set of workspaces

## Implementation challenges

I believe most implementation is straightforward -- add a separate menu / icon to handle workspaces listing, opening, and closing; add the functionality to deal with workspaces (most already existing, as "saving/loading tree style as bookmark").

Although, there are some issues to solve:
- How to determine/set the boundary of a workspace?
  - A window? Through "fake" tabs that TST already has after restoring a tree bookmark?
- How to handle window close from firefox?
- If workspace is opened in other machines, how to synchronize their changes?
  - To start with, maybe assume only one machine is managing a workspace at a time? For example, only save workspace under a machine identifier folder in bookmark?

## Additional

I guess this may be implemented as a plugin. But unfortunately I do not have any knowledge on how to do this. Maybe existing plugin developers may find this interesting?

**Comment by piroor** [comment]
It looks to be beyond the project scope of TST so it won't be a built-in feature, thus you need to implemented it as a helper addon based on TST's API: https://github.com/piroor/treestyletab/wiki/API-for-other-addons
I'm OK to add more APIs to TST if they are really required to implement the feature.
